include/drivers/qspi.h

```diff
@@ -0,0 +1,660 @@
+/**
+ * @brief No of data lines that are used for the transfer.
+ * Some flash memories supports one, two or four lines.
+ */
+#define QSPI_DATA_LINES_SINGLE                  0x00
+#define QSPI_DATA_LINES_DOUBLE                  0x01
+#define QSPI_DATA_LINES_QUAD                    0x02
+
+/**
+ * @brief QSPI mode of the transmission
+ * For detailed description see mode in qspi_config
+ */
+#define QSPI_MODE_0                             0x00
+#define QSPI_MODE_1                             0x01
+#define QSPI_MODE_2                             0x02
+#define QSPI_MODE_3                             0x03
+
+/**
+ * @brief QSPI Address configuration
+ * Length of the address field in bits.
+ * Typical flash chips support 24bit address mode
+ */
+#define QSPI_ADDRESS_MODE_8BIT                  0x00
+#define QSPI_ADDRESS_MODE_16BIT                 0x01
+#define QSPI_ADDRESS_MODE_24BIT                 0x02
+#define QSPI_ADDRESS_MODE_32BIT                 0x03
+
+/**
+ * @brief QSPI buffer structure
+ * Structure used both for TX and RX purposes.
+ *
+ * @param buf is a valid pointer to a data buffer.
+ * Can not be NULL.
+ * @param len is the length of the data to be handled.
+ * If no data to transmit/receive - pass 0.
+ */
+struct qspi_buf {
+	u8_t *buf;
+	size_t len;
+};
+
+
+/**
+ * @brief QSPI command structure
+ * Structure used for custom command usage.
+ *
+ * @param op_code is a command value (i.e 0x9F - get Jedec ID)
+ * @param tx_buf structure used for TX purposes. Can be NULL if not used.
+ * @param rx_buf structure used for RX purposes. Can be NULL if not used.
+ */
+struct qspi_cmd {
+	u8_t op_code;
+	const struct qspi_buf *tx_buf;
+	const struct qspi_buf *rx_buf;
+};
+
+
+/**
+ * @brief QSPI controller configuration structure
+ */
+struct qspi_config {
+	/* Chip Select pin used to select memory */
+	u16_t cs_pin;
+
+	/* Frequency of the QSPI bus in [Hz]*/
+	u32_t frequency;
+
+	/* Polarity, phase, Mode
+	 * 0x00: Mode 0: Data are captured on the clock rising edge and
+	 *			data are sampled on a leading edge.
+	 *			Base level of clock is 0.(CPOL=0, CPHA=0).
+	 * 0x01: Mode 1: Data are captured on the clock rising edge and
+	 *			data are sampled on a trailing edge.
+	 *			Base level of clock is 0.(CPOL=0, CPHA=1).
+	 * 0x02: Mode 2: Data are captured on the clock falling edge and
+	 *			data are sampled on a leading edge.
+	 *			Base level of clock is 1.(CPOL=1, CPHA=0).
+	 * 0x03: Mode 3: Data are captured on the clock falling edge and
+	 *			data are sampled on a trailing edge.
+	 *			Base level of clock is 1.(CPOL=1, CPHA=1).
+	 */
+	u8_t mode : 2;
+
+	/* Defines how many lines will be used for read/write operation
+	 * 0x00: One data line
+	 * 0x01: Two data lines
+	 * 0x02: Four data lines
+	 */
+	u8_t data_lines : 2;
+
+	/* Defines how many bits are used for address (8/16/24/32)
+	 * 0x00: 8 bit address field
+	 * 0x01: 16 bit address field
+	 * 0x02: 24 bit address field
+	 * 0x03: 32 bit address field
+	 */
+	u8_t address : 2;
```

Comment:
data lines and address nbr of bytes can change per cmd.
These settings should be part of a more complete `qspi_cmd` structure

Reply:
Would you be so kind and give an example? I think that the only case when you have to change number of address bits is when you have two different memory types on one QSPI bus. In that case changing number of address bits is not a problem since you have to reconfigure peripheral anyway (to change used CS pin).
Please correct me if I am wrong.

Reply:
Indeed when initializing the session with a memory, you need to perform configuration like specifying the number of lines you support and other things, but also, on a given memory, commands could have 1 to 4 bytes address length and be done one 1 to 4 lines.
On STM32 QSPI IP, this needs to be configured before sending of each command in order to avoid data corruption.

Reply:
Despite the fact that I understand your point of view, I still think that changing QSPI bus parameters should be done be `configure.` . Anyway, if we have to consider your case maybe this will do the job?
```
struct qspi_cmd {
	u8_t op_code;
	const struct qspi_buf *tx_buf;
	const struct qspi_buf *rx_buf;
        const struct qspi_cmd_conf  *conf;
};
```

where `qspi_cmd_conf` is
```
struct qspi_cmd_conf{
	u8_t data_lines : 2;
        u8_t address : 2;
};
```
But i do not think that this solution is good enough. I am curious what other interested think about it.

Reply:
IIUC, Your proposal would work on QSPI up to 1-1-4, but from performance point of view this is sufficient.

Another point, what about dummy bytes handling?
Most read commands will use 1 dummy byte (fast read (0xB) for instance).

This would not address the OSPI needs, but maybe we can keep this for a future evolution.